moby/moby
# Add http(s) proxy properties to daemon configuration (carry 42647)

### :warning: note that the format in `daemon.json` changed in a follow-up; https://github.com/moby/moby/pull/43448


The new format uses a `"proxies"` key that holds the proxies;

```json
{
  "proxies": {
    "http-proxy": "http-config",
    "https-proxy": "https-config",
    "no-proxy": "no-proxy-config"
  }
}
```

--------

carry of https://github.com/moby/moby/pull/42647
fixes https://github.com/moby/moby/issues/24758
closes https://github.com/moby/moby/pull/42647
addresses https://github.com/moby/moby/issues/40201

This allows configuring the daemon's proxy server through the daemon.json con-
figuration file or command-line flags configuration file, in addition to the
existing option (through environment variables).

Configuring environment variables on Windows to configure a service is more
complicated than on Linux, and adding alternatives for this to the daemon con-
figuration makes the configuration more transparent and easier to use.

The configuration as set through command-line flags or through the daemon.json
configuration file takes precedence over env-vars in the daemon's environment,
which allows the daemon to use a different proxy. If both command-line flags
and a daemon.json configuration option is set, an error is produced when starting
the daemon.

Note that this configuration is not "live reloadable" due to Golang's use of
`sync.Once()` for proxy configuration, which means that changing the proxy
configuration requires a restart of the daemon (reload / SIGHUP will not update
the configuration.

With this patch: (⚠️ https://github.com/moby/moby/pull/43448 changes the location of these fields to be in a "proxy-config" struct within `daemon.json`)

    cat /etc/docker/daemon.json
    {
        "http-proxy": "http://proxytest.example.com:80",
        "https-proxy": "https://proxytest.example.com:443"
    }

    docker pull busybox
    Using default tag: latest
    Error response from daemon: Get "https://registry-1.docker.io/v2/": proxyconnect tcp: dial tcp: lookup proxytest.example.com on 127.0.0.11:53: no such host

    docker build .
    Sending build context to Docker daemon  89.28MB
    Step 1/3 : FROM golang:1.16-alpine AS base
    Get "https://registry-1.docker.io/v2/": proxyconnect tcp: dial tcp: lookup proxytest.example.com on 127.0.0.11:53: no such host

Integration tests were added to test the behavior:

- verify that the configuration through all means are used (env-var,
command-line flags, damon.json), and used in the expected order of
preference.
- verify that conflicting options produce an error.
- verify that logs and error messages sanitise proxy URLs (as they may contain username / password)


```bash
make BIND_DIR=. DOCKER_GRAPHDRIVER=vfs TEST_FILTER=TestDaemonProxy binary test-integration
Running integration-test (iteration 1)
Running /go/src/github.com/docker/docker/integration/daemon (amd64.integration.daemon) flags=-test.v -test.timeout=5m  -test.run TestDaemonProxy
=== RUN   TestDaemonProxy
=== RUN   TestDaemonProxy/environment_variables
=== RUN   TestDaemonProxy/command-line_options
=== RUN   TestDaemonProxy/configuration_file
=== RUN   TestDaemonProxy/conflicting_options
=== RUN   TestDaemonProxy/reload_sanitized
--- PASS: TestDaemonProxy (6.75s)
    --- PASS: TestDaemonProxy/environment_variables (1.84s)
    --- PASS: TestDaemonProxy/command-line_options (1.84s)
    --- PASS: TestDaemonProxy/configuration_file (1.93s)
    --- PASS: TestDaemonProxy/conflicting_options (0.52s)
    --- PASS: TestDaemonProxy/reload_sanitized (0.63s)
PASS

DONE 6 tests in 6.942s
```

**- Description for the changelog**


```markdown
- Add options to the `daemon.json` configuration file and `dockerd` command-line
  to configure the daemon's proxy. With these options it is possible to configure
  http(s) proxies for the daemon through the configuration file as an alternative
  to the `HTTP_PROXY`, `HTTPS_PROXY`, and `NO_PROXY` environment variables, or to
  override the system-wide proxy configuration set in those environment variables.
```


**- A picture of a cute animal (not mandatory but encouraged)**
